[decision] Changetracking (round 3) (#4841)
ElektraInitiative/libelektra: doc/decisions/4_in_review/change_tracking.md

## Decision

> Plugins and application developer can enable changetracking via a contract.

Reviewer: I don't really see the use case for this. Important is that admins can enable change tracking via `kdb`.

Reviewer 2: I don't see it would work from the technical POV for a plugin to enable/disable changetracking.

I also agree that it is most important that it can be enabled/disabled via `kdb`, i.e. that there is a config key somewhere in `/elektra`.

Author: I don't think disabling via `/elektra` is a good way. We're talking about change tracking, NOT recording here. Some plugins might depend on changetracking, i.e. the notification plugins. 

Wouldn't it be possible that plugins export, i.e. `system:/elektra/modules/<pluginname>/requirechangetracking` in their contract?

Reviewer 2: Then you need to define what exactly "enabling changetracking" does. Also is there even a use case for disabling it? Shouldn't it just be automatically active whenever it's needed? Because than I'd say it is not about "enabling changetracking", but about "declaring that changetracking is required".

Author: Enabling changetracking means we fill the additional keyset in `struct _KDB` with deep-duped keys. If it is not enabled, the keyset will not be allocated.

Author: It's just an optimization - if no plugin requires it, why do it? Yes I probably should change the wording from "enabling" to "requiring"

> Store deep-duped returned keys in a separate KeySet, which we might be able to reuse as internal cache later on.

Reviewer: Please try to also use it as cache. If it doesn't work, simply update the decision that it didn't work. To do it as separate work later is very unlikely to be ever done.

```suggestion
Store deep-duped returned keys in a separate KeySet, which we might also use as internal cache.
```

Reviewer 2: I think this sentence needs more explanation. If "internal cache" refers to `doc/decisions/4_decided/internal_cache.md`, then the file should be linked here. Also we should mention that deep-duped means COW-copied, not fully detached copies.

Author: Yes, it refers to that.

The whole changetracking logic lives within `libelektra-kdb`.
We provide an API for developers with `libelektra-kdb`.

## Rationale

This decision meets all the constraints.

We are not altering the behaviour of already existing functions, so everything is transparent for applications using the public Elektra API.

> As changetracking is an opt-in feature, it will create zero overhead in usecases where it is not needed.

Reviewer: This contradicts with the Rationale below. (`struct _KDB` gets an additional pointer etc.).

```suggestion
As changetracking is an opt-in feature, it will create negligible overhead if not needed.
```

Reviewer 2: I think "overhead" refers to runtime here, so how about:

```suggestion
As changetracking is an opt-in feature, it will create zero runtime overhead and negligible memory overhead in usecases where it is not needed.
```

Author: What would be the difference between runtime and memory overhead?

Author: Oh you mean CPU cycles?

There is only a single implementation and storage, so no duplication for each plugin that wants change tracking.
As we are storing our own tracking data, all sequences of `kdbGet` and `kdbSet` should work.

## Implications

- Documentation for changetracking APIs needs to be written
- `struct _KDB` needs to be extended to contain the keyset for changetracking

> - Replace all custom changetracking code within Elektra with this unified implementation.

Reviewer: :+1:

>   This should also include the `keyNeedsSync` flag.

Reviewer 2: I thought we want to keep `keyNeedsSync`?

But we should mention:

```suggestion
  This should also include the `keyNeedsSync` flag.
  It also includes switching all notification plugins to use the new API.
```
